### PR TITLE
Add basic thread_t type and basic scheduler

### PR DESCRIPTION
**Pull request by dreamos82**
closing #117 and #118

### CODE REVIEW

File: src/asm/isr.s [hunk 1]
```diff
@@ -5,16 +5,18 @@
...
-	sti
-    iretq
+    save_context ; Now we can save the general purpose registers
+    mov rdi, rsp    ; Let's set the current stack pointer as a parameter of the interrupts_handler
+    cld ; Clear the direction flag
+    call interrupts_handler ; Now we call the interrupt handler
+    mov rsp, rax    ; use the returned context
+    restore_context ; We served the interrupt let's restore the previous context
+    add rsp, 16 ; We can discard the interrupt number and the error code
+    sti
```
Comment (DeanoBurrito) [inline src/asm/isr.s:18]: enabling interrupts inside of an interrupt handler is a bad idea! All your idt entries should be set with interrupt gates, which clears the interrupt flag when the iret frame is pushed onto the stack. iret then pops rflags from the stack, restoring interrupts. So no need to mess with the interrupt flags.

File: src/include/kernel/scheduling/thread.h [hunk 2]
```diff
@@ -3,25 +3,50 @@
 
 #include <stdint.h>
 #include <stddef.h>
+#include <cpu.h>
 
+#define TASK_NAME_MAX_LEN 32
 #define THREAD_NAME_MAX_LEN 32
-#define THREAD_MAX_ID (uint8_t-1)
+#define THREAD_MAX_ID (uint16_t-1)
+#define THREAD_TICKS 0x50
```
Comment (DeanoBurrito) [inline src/include/kernel/scheduling/thread.h:11]: never used?

File: src/kernel/scheduling/thread.c [hunk 3]
```diff
@@ -3,12 +3,76 @@
...
+    new_thread->execution_frame->rdi = _entry_point;
+    new_thread->execution_frame->rsi = arg;
+    new_thread->execution_frame->rflags = 0x202;
+    new_thread->execution_frame->ss = 0x10;
+    new_thread->execution_frame->cs = 0x8;
+    // We need to allocate a new stack for each thread
+    uint64_t *stack_pointer = kmalloc(THREAD_DEFAULT_STACK_SIZE);
+    
+    // The stack grow backward, so the pointer will be the end of the stack
+    new_thread->stack = stack_pointer + (THREAD_DEFAULT_STACK_SIZE);
```
Comment (DeanoBurrito) [inline src/kernel/scheduling/thread.c:30]: why the brackets ?:D

Comment (dreamos82) [reply src/kernel/scheduling/thread.c:30]: don't know probably leftover from some old tries with the stacks XD

**Pull request merged**